Sources/StreamChat/Repositories/MessageRepository_Tests.swift

```diff
@@ -0,0 +1,347 @@
+
+        var currentMessageState: LocalMessageState?
+        try database.writeSynchronously { session in
+            currentMessageState = session.message(id: id)?.localMessageState
+        }
+
+        XCTAssertEqual(currentMessageState, .sendingFailed)
+        XCTAssertEqual(result?.error, MessageRepositoryError.failedToSendMessage)
+    }
+
+    func tests_sendMessage_APISuccess() throws {
+        let id = MessageId.unique
+        try createMessage(id: id, localState: .pendingSend)
+        let expectation = self.expectation(description: "Send Message completes")
+        var result: Result<ChatMessage, MessageRepositoryError>?
+        repository.sendMessage(with: id) {
+            result = $0
+            expectation.fulfill()
+        }
+
+        wait(for: [apiClient.request_expectation], timeout: 0.1)
+
+        let payload = MessagePayload.Boxed(message: .dummy(messageId: id, authorUserId: .anonymous))
+        (apiClient.request_completion as? (Result<MessagePayload.Boxed, Error>) -> Void)?(.success(payload))
+
+        wait(for: [expectation], timeout: 0.1)
+
+        var currentMessageState: LocalMessageState?
+        try database.writeSynchronously { session in
+            currentMessageState = session.message(id: id)?.localMessageState
+        }
+
+        XCTAssertNil(currentMessageState)
+        XCTAssertNotNil(result?.value)
+    }
+
+    @discardableResult
+    private func createMessage(id: MessageId, localState: LocalMessageState) throws -> MessageDTO {
+        try database.createCurrentUser()
+        try database.createChannel(cid: cid)
+        var message: MessageDTO!
+        try database.writeSynchronously { session in
+            message = try session.createNewMessage(
+                in: self.cid,
+                text: "Message pending send",
+                pinning: nil,
+                quotedMessageId: nil,
+                isSilent: false,
+                extraData: [:]
+            )
+            message.id = id
+            message.localMessageState = localState
+        }
+        return message
+    }
+
+    private func runSendMessageAndWait(id: MessageId) -> Result<ChatMessage, MessageRepositoryError>? {
+        let expectation = self.expectation(description: "Send Message completes")
+        var result: Result<ChatMessage, MessageRepositoryError>?
+        repository.sendMessage(with: id) {
+            result = $0
+            expectation.fulfill()
+        }
+        waitForExpectations(timeout: 0.1, handler: nil)
+        return result
+    }
+
+    // MARK: saveSuccessfullySentMessage
+
+    func test_saveSuccessfullySentMessage_noChannel() {
+        let id = MessageId.unique
+        let payload = MessagePayload.dummy(messageId: id, authorUserId: .anonymous, channel: nil)
+        let message = runSaveSuccessfullySentMessageAndWait(payload: payload)
+        XCTAssertNil(message)
+    }
+
+    func test_saveSuccessfullySentMessage_channelPayload_sending() throws {
```

Comment:
The following few are very similar tests. Might be good to extract re-use the logic and only pass the local state as a parameter.

Reply:
I tried, but then the whole "script" of the test is lost. No Give-When-Then kinda way of reading it and/or the function that extracts it gets way too many parameters